### ISSUE TITLE
Render CMSPlugins as children of TextPlugin

### ISSUE THREAD

**Issue by jrief** [divio/djangocms-text-ckeditor#408]
I created a CMSPlugins, which can be added as a child to the TextPlugin. This plugin shall render something such as ``<i class="arbitrary"></i>``. The problem is, that in the CKEditor widget, this HTML never shows up.

I then tried to override the method ``icon_src()``, but without success. Why isn't it possible to wrap the rendered content into something, such as say ``<cms-plugin ...><i class="arbitrary"></i></cms-plugin>`` when visible in the editor.

Side note: Outside of the editor, the plugin instance is rendered as expected. It's just that while editing in CKEditor, plugins should be WYSIWYG as well.

**Comment by vxsx** [comment]
I believe the issue here is missing configuration. CKEditor by default doesn't provide any styles, and has specific default options. For example - it removes empty tags. You may also notice that links are not of the same color as in your website for example.

To fix this, one should configure the CKEditor plugin to use your stylesheet and not remove the empty tags, like this:

Settings: https://github.com/divio/django-cms-demo/blob/master/src/settings.py#L318-L324
js: https://github.com/divio/django-cms-explorer/blob/master/static/js/addons/ckeditor.wysiwyg.js#L73 (relevant line shows the rule for `span` tag, but it can be any)

It is also possible to explicitly disable "inline" rendering of the child plugin by setting ` text_editor_preview = False` on the plugin (e.g. https://github.com/divio/djangocms-text-ckeditor/blob/d6dd84c3b5e5574b9b0c2d19bcd15d22cdf19092/djangocms_text_ckeditor/test_app/cms_plugins.py#L9) - instead that will render the block with plugin name on it.

**Comment by jrief** [comment]
Thanks @vxsx , that was the missing part. Now it almost works for me.

One thing however, it would be great if I could configure the ``configuration['contentCss']`` by iterating over all text-enabled ``installed_plugins`` and retrieve their special TextEditor stylesheets. Otherwise this has to be done by the client through ``settings.py``, which can error-prone.

Shall I create a pull request for it?

**Comment by vxsx** [comment]
It sounds like an interesting idea, to be in control, but i have to say i'm very biased to the current way of things, even though it's [not without problems](https://github.com/divio/djangocms-text-ckeditor/issues/360).

It should be backwards compatible, as today we are usually using the site styles (it helps not only for plugins like icons but also for site specific styles - fonts, colors, button styles etc), we don't want to break that.

 @czpython @FinalAngel wdyt?

**Comment by FinalAngel** [comment]
It sounds more like magic and so far we follow the rule: just use the CKEDitor settings and it works. If then it should be something on top of it not messing with the standards. Though I don't see much use cases for special ckeditor css files per plugin.

**Comment by jrief** [comment]
@FinalAngel 
> I don't see much use cases for special ckeditor css files per plugin

I created a text-enabled CMSPlugin to render arbitrary icons inside edited text.
Into this plugin, you can upload an arbitrary font created on http://fontello.com/ . This font is shipped with some SVG, TTF, WOFF and a CSS file. My plugin serves these files through MEDIA_ROOT. Now each icon can be rendered using something such as ``<i class="icon-blah"></i>``, provided that the CSS file can be loaded.

This works perfectly when rendering the plugin, but I would also like to render the icon inside the WYSIWIG area of the CKEditor. In order to to that I have to load that CSS file into the iframe of the CKEditor. Using a configuration, as @vxsx proposed works fine, but that's static. I need a solution where I can load CSS files into the CKEditor determined during runtime.

One possible solution would be to allow each text-enabled plugin to hook into the initialization phase of CKEditor. Another would be to allow plugins to bring their own CSS files. Maybe there are other and better solutions.

**Comment by jrief** [comment]
Finally I found a working solution, which can dynamically load stylesheets and other assets from MEDIA_ROOT. The only missing part is a small fix on the **djangocms-text-ckeditor** side.

Therefore this issue can be closed via #409.

**Comment by FinalAngel** [comment]
Nice, how would you load assets dynamically?

**Comment by jrief** [comment]
It works by adding 

```
CKEDITOR_SETTINGS = {
    ...
    'stylesSet': format_lazy('default:{}', reverse_lazy('admin:some_view')),
}
```
where "some_view" renders a Javascript code snippet.

The important part here is, that ``json.dumps()`` must be able to handle lazy objects, because of ``reverse_lazy``. This is fixed by #409 and btw. is good practice anyway, because the ``DjangoJSONEncoder`` also handles dates, decimals and promises.

**Comment by czpython** [comment]
Fixed by https://github.com/divio/djangocms-text-ckeditor/pull/409
